docs/proposals/04-new-core-gardener-cloud-apis.md

```diff
@@ -0,0 +1,727 @@
+# seedSelector:
+#   matchLabels:
+#     foo: bar
+  kubernetes:
+    versions:
+    - version: 1.12.1
+    - version: 1.11.0
+    - version: 1.10.6
+    - version: 1.10.5
+      expirationDate: 2020-04-05T01:02:03Z # optional
+  backup: # See https://github.com/gardener/gardener/blob/master/docs/proposals/02-backupinfra.md.
+    type: <some-provider-name> # {aws,azure,gcp,...}
+  # region: eu-west-1
+    secretRef:
+      name: backup-secret
+      namespace: garden
+  machineTypes:
+  - name: m5.large
+    cpu: "2"
+    gpu: "0"
+    memory: 8Gi
+  # storage: 20Gi # optional (not needed in every environment, may only be specified if no volumeTypes have been specified)
+    usable: true
+  volume:
+    minimumSize: 20Gi
+    types: # optional (not needed in every environment, may only be specified if no machineType has a `storage` field)
+    - name: gp2
+      class: standard
+    - name: io1
+      class: premium
+  machineImages:
+  - name: coreos
+    versions:
+    - version: 2023.5.0
+    - version: 1967.5.0
+      expirationDate: 2020-04-05T08:00:00Z
+  - name: ubuntu
+    versions:
+    - version: 18.04.201906170
+  zones: # optional (not needed in every environment)
+  - region: europe-central-1
+    zones:
+    - name: europe-central-1a
+    - name: europe-central-1b
+    - name: europe-central-1c
+    # unavailableMachineTypes: # optional, list of machine types defined above that are not available in this zone
+    # - name: m5.large
+    # unavailableVolumeTypes: # optional, list of volume types defined above that are not available in this zone
+    # - name: io1
+# CA bundle that will be installed onto every shoot machine that is using this provider profile.
+# caBundle: |
+#   -----BEGIN CERTIFICATE-----
+#   ...
+#   -----END CERTIFICATE-----
+  providerConfig:
+    <some-provider-specific-controlplane-config>
+    # We don't have concrete examples for every existing provider yet, but these are the proposals:
+    #
+    # Example for Alicloud:
+    #
+    # apiVersion: alicloud.provider.extensions.gardener.cloud/v1alpha1
+    # kind: ProviderProfileConfig
+    # machineImages:
+    # - name: coreos
+    #   version: 2023.5.0
+    #   id: coreos_2023_4_0_64_30G_alibase_20190319.vhd
+    #
+    #
+    # Example for AWS:
+    #
+    # apiVersion: aws.provider.extensions.gardener.cloud/v1alpha1
+    # kind: ProviderProfileConfig
+    # machineImages:
+    # - name: coreos
+    #   version: 1967.5.0
+    #   regions:
+    #   - name: europe-central-1
+    #     ami: ami-0f46c2ed46d8157aa
+    #
+    #
+    # Example for Azure:
+    #
+    # apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
+    # kind: ProviderProfileConfig
+    # machineImages:
+    # - name: coreos
+    #   version: 1967.5.0
+    #   publisher: CoreOS
+    #   offer: CoreOS
+    #   sku: Stable
+    # countFaultDomains:
+    # - region: westeurope
+    #   count: 2
+    # countUpdateDomains:
+    # - region: westeurope
+    #   count: 5
```

Comment:
As we plan to support also Zones beside AvailabilitySets for Azure, we should maybe move to a `regions` list here.
AvSets will be possible in every region (means count update and failure domains needs to be always available). Zones are only available in dedicated regions. Zones names are numerically arranged.
```yaml
regions:
- name: westeurope
  countUpdateDomains: 5
  countFailureDomains: 2
  zones:
  - 1
  - 2
  - 3
- name: northeurope
  countUpdateDomains: 5
  countFailureDomains: 2
```

Reply:
There is already https://github.com/gardener/gardener/pull/1284/files#diff-bf2774d9954baab517306db45a5b80bbR116-R125, not sure if you saw this. I'll change this to

```yaml
  regions:
  - name: europe-central-1
    zones: # optional (not needed in every environment)
    - name: europe-central-1a
    - name: europe-central-1b
    - name: europe-central-1c
```

You can define the azure zones there. The `countUpdateDomains` and `countFailureDomains` are very special, so I suggest to keep it as it is (we should not duplicate the mentioned generic regions listed above).
WDYT?

Reply:
Ah sorry missed that. Yes I see the advantage to stay with the generic `regions` list in the `ProviderProfile` also for Azure and to move the AvailabilitySet specific things to the `ProviderProfileConfig`.

On the other hand side this means somehow double maintenance of the region/zone configuration. For example we need for the region `westeurope` an entry in the generic region list and another entry in the Azure `ProviderProfileConfig`. This double maintenance applies also to the additional field like `unavailableMachineTypes` and `unavailableVolumeTypes`.

I have not really an option on what's the best way to go.

Reply:
That's the price of extensibility :-)

Reply:
One remark to the fields `unavailableMachineTypes` and `unavailableMachineTypes` in the `region` list for Azure.
Some machineTypes e.g. m-series (I'm not sure for volumes) are not available in all zones of a region and the zone numbers/names are not stable of Azure. Don't know if this applies also to other providers.

**Example:**
Region westeurope consists of 3 zones. 2 of these zones have a special machine type, the last one not.
In account/subscription A points zone number `1` to a zone which have the special machine type.
In account/subscription B points zone number  `1` to a zone which **don't** have the special machine type.

That means we can't use the  field `unavailableMachineTypes` for Azure, but this is maybe not an issue as these fields are optional. That means the user needs to know the zone assignments for its subscription and need to select the correct zones in the Shoot Spec `.spec.provider.workers[].zones[]`

Reply:
Hm, good point, AFAIK the same is true for AWS (and also Alicloud (@jia-jerry)?). Does it then make sense to not have this field at all? /cc @vlerenc

Reply:
I think field `unavailableMachineTypes` is neccessary. At least for AliCloud, it is needed to maintain the relationship between zones and supported machine types.

Whatsmore, I would suggest that our cloud provider extension should validate supported machine types for zones when ProviderProfile is chanage. Even periodic validation would be better. Machine type and zone mis-match issue occurred several times for shoot on AliCloud.

Reply:
So, you say in Alicloud it's not the case that zones are mapped to the same technical zones in all accounts, not like in AWS/Azure?

Reply:
Maybe it has. I remember some one told me some machine type only be available for some special account. I need to double confirm with AliCloud guys. From this point of view, * unavailableMachineTypes* maybe not enough.

Reply:
Usually, some machine types are only for special contract customers. Maybe we should mark this special machine type warning only special account can consume it.

Reply:
Well, then I don't think it makes sense to have this field at all. Marking some machine types for special accounts is also unmaintainable IMHO. Shall we drop it?

Reply:
But for normal account in AliCloud, not all zones share same machine types. This might be special case in AliCloud.

Reply:
For special machine types, I would suggest add a flag `vip` flag (optional). I don't know whethe it is applicable for azure or aws case.
```yaml
  machineTypes:
  - name: m5.large
    cpu: "2"
    gpu: "0"
    memory: 8Gi
  # storage: 20Gi # optional (not needed in every environment, may only be specified if no volumeTypes have been specified)
    usable: true
    vip: true
```

Reply:
I don't understand. You say the zones are also mapped differently per account in Alicloud. Then it doesn't make sense to maintain such a list on our side because we cannot do it for all accounts. Hence, I'd drop the field completely because it cannot work reliably.

Regarding the `vip`, I don't see how this should work and I think it should be discussed outside of this PR.

Reply:
I didn't say there is a technical zone, which is not supported in AliCloud as far as I know. But for AliCloud I just confirmed, there is some special machine types. These types can only be consumed by some special accounts. Anyway, this case we could talk out side the PR. Please be noted, machine types and zone types are static.

For this PR, for normal account, we must care about following case.  Refer https://github.com/gardener/gardener/blob/master/example/30-cloudprofile-alicloud.yaml#L27-L33:
Machine type `ecs.sn2ne.large` is availble for zone `cn-beijing-f`, but not available in zone `cn-beijing-a`.
Without field `unavailableMachineTypes`, we need to find an optional solution to maintain such relationship.

Reply:
OK, then I misunderstood. If every zone maps to the same underlying infrastructure independent of the account then let's keep it. Thank you for clarifying.